glib/src/object.rs

```diff
@@ -3267,6 +3294,86 @@ wrapper! {
     }
 }
 
+struct WeakRefNotifyData {
+    ptr: *mut GObject,
+    callback: Box<dyn Fn()>,
+    should_drop: AtomicBool,
+}
+
+impl WeakRefNotifyData {
+    pub fn new<F: Fn() + 'static>(ptr: *mut GObject, f: F) -> ManuallyDrop<Pin<Box<Self>>> {
+        ManuallyDrop::new(Box::pin(WeakRefNotifyData {
+            ptr,
+            callback: Box::new(f),
+            should_drop: AtomicBool::new(false),
+        }))
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Checks whether the data should be dropped and marks it to be dropped
+    pub fn should_drop(&self) -> bool {
+        self.should_drop.swap(true, Ordering::SeqCst)
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// A handle to disconnect a weak ref notify closure.
+pub struct WeakRefNotify(
+    ManuallyDrop<Pin<Box<WeakRefNotifyData>>>,
+    gobject_ffi::GWeakNotify,
+);
+
+impl WeakRefNotify {
+    pub(crate) fn new<T: ObjectType, F: Fn() + 'static>(obj: &T, f: F) -> WeakRefNotify {
+        unsafe extern "C" fn notify_func<F: Fn()>(
+            data: ffi::gpointer,
+            _obj: *mut gobject_ffi::GObject,
+        ) {
+            let data = &mut *(data as *mut WeakRefNotifyData);
+
+            (data.callback)();
+
+            // Drop the data when the `WeakRefNotify` handle was already dropped
+            if data.should_drop() {
+                std::ptr::drop_in_place(data);
+            }
+        }
+
+        unsafe {
+            let obj_ptr = obj.as_object_ref().to_glib_none().0;
+            let mut data = WeakRefNotifyData::new(obj_ptr, f);
+
+            let data_ptr: *mut _ = Pin::as_mut(&mut data).get_unchecked_mut();
+            gobject_ffi::g_object_weak_ref((*data).ptr, Some(notify_func::<F>), data_ptr as *mut _);
+
+            WeakRefNotify(data, Some(notify_func::<F>))
+        }
+    }
+
+    #[doc(alias = "g_object_weak_unref")]
+    pub fn disconnect(self) {
+        unsafe {
+            let data = &*(self.0);
+            let data_ptr: *const _ = Pin::as_ref(&self.0).get_ref();
+
+            if !data.should_drop() {
+                gobject_ffi::g_object_weak_unref(data.ptr, self.1, data_ptr as *mut _);
```

Comment:
This will do bad things if the object was destroyed in the meantime. Your usage of the atomic bool is not preventing that

Maybe we just can't provide a way to disconnect safely

Reply:
> This will do bad things if the object was destroyed in the meantime. Your usage of the atomic bool is not preventing that

Are you sure? Once the callback is called `should_drop` is set to `true` and `gobject_ffi::g_object_weak_unref()` isn't called anymore.

Reply:
Thread A can see `should_drop == false`, then thread B calls the notify callback, then thread A calls `g_object_weak_unref`.

Reply:
right. I think we should hold instead of the `mut * GObject` a `WeakRef` to the object and then try to `upgrade()`  it before we call `gobject_ffi::g_object_weak_unref()` this way the `ref_count` is definitely >1 and the callback won't be called from any thread.

Reply:
That probably works, yes